client/tracing/src/logging/mod.rs

```diff
@@ -0,0 +1,531 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2021 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Substrate logging library.
+//!
+//! This crate uses tokio's [tracing](https://github.com/tokio-rs/tracing/) library for logging.
+
+#![warn(missing_docs)]
+
+mod directives;
+mod event_format;
+mod layers;
+
+pub use directives::*;
+pub use sc_tracing_proc_macro::*;
+
+use sc_telemetry::{ExtTransport, TelemetryWorker};
+use std::io;
+use tracing::Subscriber;
+use tracing_subscriber::{
+	fmt::time::ChronoLocal,
+	fmt::{
+		format, FormatEvent, FormatFields, Formatter, Layer as FmtLayer, MakeWriter,
+		SubscriberBuilder,
+	},
+	layer::{self, SubscriberExt},
+	registry::LookupSpan,
+	EnvFilter, FmtSubscriber, Layer, Registry,
+};
+
+pub use event_format::*;
+pub use layers::*;
+
+/// Logging Result typedef.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Logging errors.
+#[derive(Debug, thiserror::Error)]
+#[allow(missing_docs)]
+#[non_exhaustive]
+pub enum Error {
+	#[error(transparent)]
```

Comment:
I copy-pasted things around. Maybe I'm not using this right. What should I put instead? `#[error("IO error")]`?

Reply:
I'm hazy on the details myself but looking at [the docs](https://github.com/dtolnay/thiserror/commit/1cce98b133e296631a04617ed7f043ac7cfdeea8) it seems like you'd use `transparent` if you want to delegate things to the source error. My question then becomes: if you don't want this `Error` enum to process the underlying errors at all, then maybe it's not needed. The rule of thumb for introducing your own error type is that you actually want to do something special with it (even as simple as providing your own error messages); if they're all transparent and delegate then what's the point?

Reply:
You still have one thing: you have a new `Error` type dedicated to this module. So now you know that if that error raises somewhere in the client: then you know it comes from the logging initialization because it's an `sc_tracing::logging::Error`, 

It's actually more important for the `Error` type that will wrap this error (sc-service::Error I believe) to put a clear error message on that:

```
#[error("Could not initialize logging")]
LoggingInitializationError(sc_tracing::logging::Error),
```

The alternative to that would be to use `anyhow` or a `Box<dyn Error>` but in that case you lose the information that this error happened in sc_tracing::logging.

In my own modules I didn't use `transparent`, I used `#[error("IO error")]` (for example). The only thing that changed is that the message is prefixed by "IO error" but in the end what really matter to me was the grouping `enum`.

If you are not sure either maybe we should ask tomaka for his enlightenment on that.

Reply:
Let's stick `#[error(transparent)]` outside the item then:

```
#[error(transparent)]
pub enum Error {
	IoError(#[from] io::Error),
	SetGlobalDefaultError(#[from] tracing::subscriber::SetGlobalDefaultError),
	DirectiveParseError(#[from] tracing_subscriber::filter::ParseError),
	SetLoggerError(#[from] tracing_log::log_tracer::SetLoggerError),
}
```

Reply:
Ok done